### PR TITLE
Add ext. methods to replace LINQ in generated code

### PR DESCRIPTION
**Pull request by kasperk81**
part of https://github.com/microsoft/kiota/issues/4732

### CODE REVIEW

File: src/extensions/IEnumerableExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+using System.Collections.Generic;
+
+namespace Microsoft.Kiota.Abstractions.Extensions
+{
+    /// <summary>
+    /// Extension methods for <see cref="IEnumerable"/>
+    /// </summary>
+    public static class IEnumerableExtensions
+    {
+        public static List<T> AsList<T>(this IEnumerable<T> e) => new List<T>(e);
```
Comment (baywet) [inline src/extensions/IEnumerableExtensions.cs:16]: Should protect against the source being null

File: src/extensions/IEnumerableExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        public static T[] AsArray<T>(this IEnumerable<T> e)
+        {
+            if (e is ICollection<T> collection)
```
Comment (baywet) [inline src/extensions/IEnumerableExtensions.cs:20]: Should protect against the source being null

File: src/extensions/IEnumerableExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            {
```
Comment (baywet) [inline src/extensions/IEnumerableExtensions.cs:21]: If the source is already an array, we should probably just cast and return that.

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for starting this. A couple of comments. 
+ Version bump
+ Changelog entry
+ Unit tests

File: src/extensions/IEnumerableExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        /// <summary>
+        /// Converts an <see cref="IEnumerable{T}"/> to a <see cref="List{T}"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of elements in the collection.</typeparam>
+        /// <param name="e">The enumerable to convert.</param>
+        /// <returns>A <see cref="List{T}"/> containing the elements of the enumerable.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the input enumerable is null.</exception>
+        public static List<T> AsList<T>(this IEnumerable<T> e)
+        {
+            if (e is null) throw new ArgumentNullException(nameof(e), "Input collection cannot be null.");
```
Comment (baywet) [inline src/extensions/IEnumerableExtensions.cs:25]: what do you think about returning null instead to preserve the behaviour of what we have today with null propagation?

File: src/extensions/IEnumerableExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            T[]? result = null;
+
+            if (e is ICollection<T> collection)
+            {
+                // Allocate an array with the exact size
+#if NET5_0_OR_GREATER
+                result = GC.AllocateUninitializedArray<T>(collection.Count);
+#else
+                result = new T[collection.Count];
+#endif
```
Comment (baywet) [inline src/extensions/IEnumerableExtensions.cs:53]: could we avoid the duplication by having a small method that accepts the length as a parameter?

**Review by baywet — APPROVED** [review]
Thanks for making the changes! @andrueastman for final review and merge

**Review by andrueastman — APPROVED** [review]
👍🏼

**Pull request merged**